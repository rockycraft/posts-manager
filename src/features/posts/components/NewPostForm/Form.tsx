import { memo } from "react";
import { FormProps } from "@/types/FormProps";
import { Button, HStack, Input } from "@chakra-ui/react";
import { Post } from "@/types/Post";

function Form({ onSubmit, register }: FormProps<Omit<Post, "id">>) {
	return (
		<form role="form" onSubmit={onSubmit}>
			<HStack>
				<Input
					data-testid="name-input"
					placeholder="Nombre"
					type="text"
					{...register("name")}
					variant="subtle"
				/>
				<Input
					data-testid="description-input"
					placeholder="DescripciÃ³n"
					type="text"
					{...register("description")}
					variant="subtle"
				/>
				<Button data-testid="submit-button" type="submit" colorPalette="blue">
					Crear
				</Button>
			</HStack>
		</form>
	);
}

export default memo(Form);
