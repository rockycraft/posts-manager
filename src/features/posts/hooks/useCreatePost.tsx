import type { State } from "@/types/State";
import { useCallback, useState } from "react";
import { useDispatch } from "react-redux";
import { createPost } from "../slices/postsSlice";
import { postPost } from "../services/postPost";
import type { Post } from "@/types/Post";

export function useCreatePost() {
	const [state, setState] = useState<State>("idle");
	const dispatch = useDispatch();

	const create = useCallback(
		async (post: Omit<Post, "id">) => {
			try {
				setState("loading");
				const response = await postPost(post);
				dispatch(createPost(response));
				setState("success");
			} catch (error) {
				setState("error");
			}
		},
		[dispatch],
	);

	return { create, state };
}
