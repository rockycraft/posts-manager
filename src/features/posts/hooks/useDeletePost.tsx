import type { State } from "@/types/State";
import { useState } from "react";
import { deletePost as deletePostService } from "../services/deletePost";
import { useDispatch } from "react-redux";
import { removePost } from "../slices/postsSlice";
import { Post } from "@/types/Post";

export function useDeletePost() {
	const [state, setState] = useState<State>("idle");
	const dispatch = useDispatch();

	async function deletePost(id: Post["id"]) {
		try {
			setState("loading");
			//@ts-ignore
			const response = await deletePostService(id);
			dispatch(removePost(response.id));
			setState("success");
		} catch (error) {
			console.error(error);
			setState("error");
		}
	}

	return { deletePost, state };
}
