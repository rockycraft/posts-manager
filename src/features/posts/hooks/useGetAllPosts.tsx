import type { RootState } from "@/store";
import { useCallback, useEffect, useRef, useState } from "react";
import { shallowEqual, useDispatch, useSelector } from "react-redux";
import { getPosts } from "../services/getPosts";
import { addPosts } from "../slices/postsSlice";
import type { State } from "@/types/State";

export function useGetAllPosts() {
	const [state, setState] = useState<State>("idle");
	const initialized = useRef(false);
	const posts = useSelector(
		(state: RootState) => state.posts.posts,
		shallowEqual,
	);
	const dispatch = useDispatch();

	const getAllPosts = useCallback(async () => {
		try {
			setState("loading");
			const response = await getPosts();
			dispatch(addPosts(response));
			setState("success");
		} catch (error) {
			setState("error");
			console.error(error);
		}
	}, [dispatch]);

	useEffect(() => {
		if (!initialized.current) {
			initialized.current = true;

			getAllPosts();
		}
	}, [getAllPosts]);

	return { data: posts, state, getAllPosts };
}
