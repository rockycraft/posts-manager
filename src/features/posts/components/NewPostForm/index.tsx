import type { z } from "zod";
import { postSchema } from "@/types/Post";
import { type SubmitHandler, useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import { useCreatePost } from "../../hooks/useCreatePost";
import { memo, useCallback, useEffect } from "react";
import Form from "./Form";
import { Toaster, toaster } from "@/components/ui/toaster";

const newPostSchema = postSchema.omit({ id: true });

type NewPostInput = z.infer<typeof newPostSchema>;

function NewPostForm() {
	const { register, handleSubmit, reset } = useForm<NewPostInput>({
		resolver: zodResolver(newPostSchema),
	});
	const { create, state } = useCreatePost();

	useEffect(() => {
		if (state === "error") {
			toaster.create({
				title: "Ocurri√≥ un error al crear el post",
				type: "error",
			});
		}
	}, [state]);

	const onSubmit = useCallback<SubmitHandler<NewPostInput>>(
		(data) => {
			create(data);
			reset();
		},
		[create, reset],
	);

	return (
		<>
			<Form onSubmit={handleSubmit(onSubmit)} register={register} />
			<Toaster />
		</>
	);
}

export default memo(NewPostForm);
