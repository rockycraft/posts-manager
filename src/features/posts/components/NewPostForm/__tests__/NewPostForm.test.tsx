import "@testing-library/jest-dom";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { beforeEach, describe, expect, it, vi } from "vitest";
import NewPostForm from "../index";
import { useCreatePost } from "../../../hooks/useCreatePost";
import { toaster } from "@/components/ui/toaster";
import { Provider } from "@/components/ui/provider";

Object.defineProperty(window, "matchMedia", {
	writable: true,
	value: vi.fn().mockImplementation((query) => ({
		matches: false,
		media: query,
		onchange: null,
		addListener: vi.fn(), // deprecated
		removeListener: vi.fn(), // deprecated
		addEventListener: vi.fn(),
		removeEventListener: vi.fn(),
		dispatchEvent: vi.fn(),
	})),
});

vi.mock("../../../hooks/useCreatePost");
vi.mock("@/components/ui/toaster", () => ({
	Toaster: () => <div>Toaster</div>,
	toaster: {
		create: vi.fn(),
	},
}));

describe("NewPostForm component", () => {
	const mockCreate = vi.fn();
	const mockReset = vi.fn();

	beforeEach(() => {
		vi.clearAllMocks();
		// @ts-ignore
		useCreatePost.mockReturnValue({ create: mockCreate, state: "idle" });
	});

	it("should take a snapshot", () => {
		const { asFragment } = render(
			<Provider>
				<NewPostForm />
			</Provider>,
		);
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render the from with inputs and button", () => {
		render(
			<Provider>
				<NewPostForm />
			</Provider>,
		);

		expect(screen.getByTestId("name-input")).toBeInTheDocument();
		expect(screen.getByTestId("description-input")).toBeInTheDocument();
		expect(screen.getByTestId("submit-button")).toBeInTheDocument();
	});

	it("should call create and reset on form submit", () => {
		render(
			<Provider>
				<NewPostForm />
			</Provider>,
		);

		fireEvent.input(screen.getByPlaceholderText("Nombre"), {
			target: { value: "Test Name" },
		});
		fireEvent.input(screen.getByPlaceholderText("Descripción"), {
			target: { value: "Test Description" },
		});
		fireEvent.submit(screen.getByRole("form"));

		waitFor(() => {
			expect(mockCreate).toHaveBeenCalledTimes(1);
			expect(mockCreate).toHaveBeenCalledWith({
				name: "Test Name",
				description: "Test Description",
			});
			expect(mockReset).toHaveBeenCalledTimes(1);
		});
	});

	it("should show error toaster when state is error", () => {
		//@ts-ignore
		useCreatePost.mockReturnValue({
			create: mockCreate,
			state: "error",
		});

		render(
			<Provider>
				<NewPostForm />
			</Provider>,
		);

		waitFor(() => {
			expect(toaster.create).toHaveBeenCalledWith({
				title: "Ocurrió un error al crear el post",
				type: "error",
			});
		});
	});
});
