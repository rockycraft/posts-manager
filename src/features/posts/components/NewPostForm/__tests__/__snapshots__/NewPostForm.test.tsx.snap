// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewPostForm component > should take a snapshot 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <form
    role="form"
  >
    <div
      class="chakra-stack css-1igwmid"
    >
      <input
        class="chakra-input css-1hens3x"
        data-testid="name-input"
        name="name"
        placeholder="Nombre"
        type="text"
      />
      <input
        class="chakra-input css-1hens3x"
        data-testid="description-input"
        name="description"
        placeholder="DescripciÃ³n"
        type="text"
      />
      <button
        class="chakra-button css-4hnymn"
        data-testid="submit-button"
        type="submit"
      >
        Crear
      </button>
    </div>
  </form>
  <div>
    Toaster
  </div>
</DocumentFragment>
`;
